src/stream.rs

```diff
@@ -0,0 +1,301 @@
+                            self.recv_window.set(self.config.receive_window as usize);
+                        }
+                        continue
+                    }
+                    Some(Item::WindowUpdate(n)) => {
+                        trace!("[{}] received window update: {}", self.id, n);
+                        self.send_window = self.send_window.checked_add(n).unwrap_or(u32::MAX);
+                        if let Some(task) = self.writer_task.take() {
+                            trace!("[{}] notifying writer task", self.id);
+                            task.notify()
+                        }
+                        continue
+                    }
+                    Some(Item::Finish) => {
+                        trace!("[{}] received finish", self.id);
+                        if self.state == State::SendClosed {
+                            self.state = State::Closed;
+                            return Async::Ready(())
+                        } else {
+                            self.state = State::RecvClosed
+                        }
+                        continue
+                    }
+                    Some(Item::Reset) => {
+                        trace!("[{}] received reset", self.id);
+                        self.state = State::Closed;
+                        return Async::Ready(())
+                    }
+                    None => {
+                        trace!("[{}] receiver returned None", self.id);
+                        self.state = State::Closed;
+                        return Async::Ready(())
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl io::Read for Stream {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.state == State::Closed || self.state == State::RecvClosed {
+            return Ok(0)
+        }
+        if self.poll_receiver().is_ready() {
+            return Ok(0)
+        }
+        if self.buffer.is_empty() {
+            return Err(io::Error::new(io::ErrorKind::WouldBlock, "not ready"))
+        }
+        let n = min(buf.len(), self.buffer.len());
+        (&mut buf[0..n]).copy_from_slice(&self.buffer.split_to(n));
+        Ok(n)
+    }
+}
+
+impl AsyncRead for Stream { }
+
+impl io::Write for Stream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.state == State::Closed || self.state == State::SendClosed {
+            return Err(io::Error::new(io::ErrorKind::Other, "connection closed"))
+        }
+
+        self.poll_receiver();
+
+        if self.state == State::Closed {
+            return Ok(0)
+        }
+
+        if self.send_window == 0 {
+            trace!("[{}] write: send window exhausted", self.id);
+            self.writer_task = Some(task::current());
+            return Err(io::Error::new(io::ErrorKind::WouldBlock, "window empty"))
+        }
+
+        let len = min(buf.len(), self.send_window as usize);
+        self.send_window -= len as u32;
```

Comment:
`Body::from_bytes` checks that `len <= u32::max()`, but that happens *after* this line where you do `len as u32`.
There's no chance that we try to send 4 GB of data I think, but I think the code is a bit absurd as it is.
Maybe do `Body::from_bytes(...).expect(...)` instead?

Reply:
`self.send_window` is a `u32` so `len` can not exceed `u32::MAX`.